### PR TITLE
GraalJs based Sequence handling with Configurable Engine Switch

### PR DESCRIPTION
**Pull request by shanggeeth**
## Goals

- Adaptive Authentication templates works with GraalJs
- Script engines can be switched before Identity Server Start Up

### Related Pull Requests
 - [wso2/carbon-identity-framework](https://github.com/wso2/carbon-identity-framework/pull/3444)
 - [wso2/product-is](https://github.com/wso2/product-is/pull/11447)

### CODE REVIEW

File: components/org.wso2.carbon.identity.conditional.auth.functions.acr/pom.xml [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+                        <Private-Package>
+                            org.wso2.carbon.identity.conditional.auth.functions.acr.internal
+                        </Private-Package>
+                        <Export-Package>
+                            !org.wso2.carbon.identity.conditional.auth.functions.acr.internal,
+                            org.wso2.carbon.identity.conditional.auth.functions.acr.*
+                        </Export-Package>
+                        <Import-Package>
+                            org.apache.commons.logging,
+<!--                            org.apache.http.entity,-->
```
Comment (senthalan) [inline components/org.wso2.carbon.identity.conditional.auth.functions.acr/pom.xml:77]: any specific reason to have this commented ?

File: components/org.wso2.carbon.identity.conditional.auth.functions.acr/src/main/java/org/wso2/carbon/identity/conditional/auth/functions/acr/SelectAcrFromFunction.java [hunk 2]
```diff
@@ -0,0 +1,84 @@
+package org.wso2.carbon.identity.conditional.auth.functions.acr;
```
Comment (senthalan) [inline components/org.wso2.carbon.identity.conditional.auth.functions.acr/src/main/java/org/wso2/carbon/identity/conditional/auth/functions/acr/SelectAcrFromFunction.java:1]: license missing

File: components/org.wso2.carbon.identity.conditional.auth.functions.acr/src/main/java/org/wso2/carbon/identity/conditional/auth/functions/acr/SelectAcrFromFunction.java [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+import org.wso2.carbon.identity.conditional.auth.functions.acr.internal.AcrFunctionsServiceComponent;
+
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Map;
+import java.util.TreeMap;
+
+public class SelectAcrFromFunction implements SelectOneFunction {
```
Comment (senthalan) [inline components/org.wso2.carbon.identity.conditional.auth.functions.acr/src/main/java/org/wso2/carbon/identity/conditional/auth/functions/acr/SelectAcrFromFunction.java:16]: class doc missing

File: components/org.wso2.carbon.identity.conditional.auth.functions.acr/src/main/java/org/wso2/carbon/identity/conditional/auth/functions/acr/SelectAcrFromFunction.java [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+
+    private static final Log log = LogFactory.getLog(org.wso2.carbon.identity.conditional.auth.functions.acr.SelectAcrFromFunction.class);
+
+    public String evaluate(JsAuthenticationContext context,Object... possibleOutcomesObj) {
```
Comment (senthalan) [inline components/org.wso2.carbon.identity.conditional.auth.functions.acr/src/main/java/org/wso2/carbon/identity/conditional/auth/functions/acr/SelectAcrFromFunction.java:20]: add override annotation

File: components/org.wso2.carbon.identity.conditional.auth.functions.acr/src/main/java/org/wso2/carbon/identity/conditional/auth/functions/acr/SelectAcrFromFunction.java [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+                (Comparator<Integer>) (o1, o2) -> o2.compareTo(o1)));
+        String acrSelected = null;
+
+        for (String acrChecked : acrListRequested) {
+            for (int x = 0; x < possibleOutcomes.length; x++) {
+                String outcomeToTest = possibleOutcomes[x];
+                if (outcomeToTest.equals(acrChecked)) {
+                    if (log.isDebugEnabled()) {
+                        log.debug("Reassigning Best Match for the outcome : " + outcomeToTest + " with priority : " +
+                                x+1);
```
Comment (senthalan) [inline components/org.wso2.carbon.identity.conditional.auth.functions.acr/src/main/java/org/wso2/carbon/identity/conditional/auth/functions/acr/SelectAcrFromFunction.java:72]: possible format issue

File: components/org.wso2.carbon.identity.conditional.auth.functions.acr/src/main/java/org/wso2/carbon/identity/conditional/auth/functions/acr/SelectOneFunction.java [hunk 6]
```diff
@@ -0,0 +1,8 @@
+package org.wso2.carbon.identity.conditional.auth.functions.acr;
```
Comment (senthalan) [inline components/org.wso2.carbon.identity.conditional.auth.functions.acr/src/main/java/org/wso2/carbon/identity/conditional/auth/functions/acr/SelectOneFunction.java:1]: license missing

File: components/org.wso2.carbon.identity.conditional.auth.functions.acr/src/main/java/org/wso2/carbon/identity/conditional/auth/functions/acr/internal/AcrFunctionsServiceHolder.java [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+ *  specific language governing permissions and limitations
+ *  under the License.
+ */
+package org.wso2.carbon.identity.conditional.auth.functions.acr.internal;
+
+import org.wso2.carbon.identity.application.authentication.framework.JsFunctionRegistry;
+import org.wso2.carbon.registry.core.service.RegistryService;
+import org.wso2.carbon.user.core.service.RealmService;
+
+public class AcrFunctionsServiceHolder {
```
Comment (senthalan) [inline components/org.wso2.carbon.identity.conditional.auth.functions.acr/src/main/java/org/wso2/carbon/identity/conditional/auth/functions/acr/internal/AcrFunctionsServiceHolder.java:24]: Doc comment missing

File: components/org.wso2.carbon.identity.conditional.auth.functions.acr/src/test/resources/testng.xml [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+            <class name="org.wso2.carbon.identity.conditional.auth.functions.acr.HandleAcrTest"/>
+        </classes>
+    </test>
+    <test name="send-email-graaljs" parallel="false">
+        <parameter name="scriptEngine" value="graaljs"/>
+        <classes>
+            <class name="org.wso2.carbon.identity.conditional.auth.functions.acr.HandleAcrTest"/>
+        </classes>
+    </test>
+</suite>
```
Comment (senthalan) [inline components/org.wso2.carbon.identity.conditional.auth.functions.acr/src/test/resources/testng.xml:31]: new line in end of line

File: components/org.wso2.carbon.identity.conditional.auth.functions.analytics/src/main/java/org/wso2/carbon/identity/conditional/auth/functions/analytics/CallAnalyticsFunctionImpl.java [hunk 9]
```diff
@@ -60,8 +62,11 @@
     private static final String PARAM_INPUT_STREAM = "InputStream";
 
     @Override
-    public void callAnalytics(Map<String, String> metadata,
-                              Map<String, Object> payloadData, Map<String, Object> eventHandlers) {
+    public void callAnalytics(Map<String, String> metadataJsObject,
+                              Map<String, Object> payloadDataJsObject, Map<String, Object> eventHandlers) {
+
+        Map<String, String> metadata = new HashMap<>(metadataJsObject);
+        Map<String, Object> payloadData = new HashMap<>(payloadDataJsObject);
```
Comment (senthalan) [inline components/org.wso2.carbon.identity.conditional.auth.functions.analytics/src/main/java/org/wso2/carbon/identity/conditional/auth/functions/analytics/CallAnalyticsFunctionImpl.java:69]: What is the requirement for this initialization?

File: components/org.wso2.carbon.identity.conditional.auth.functions.analytics/src/test/resources/testng.xml [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+            <class name="org.wso2.carbon.identity.conditional.auth.functions.analytics.CallAnalyticsFunctionImplTest"/>
+        </classes>
+    </test>
+    <test name="call-analytics-graaljs" parallel="false">
+        <parameter name="scriptEngine" value="graaljs"/>
+        <classes>
+            <class name="org.wso2.carbon.identity.conditional.auth.functions.analytics.CallAnalyticsFunctionImplTest"/>
+        </classes>
+    </test>
+</suite>
```
Comment (senthalan) [inline components/org.wso2.carbon.identity.conditional.auth.functions.analytics/src/test/resources/testng.xml:31]: new line.

File: components/org.wso2.carbon.identity.conditional.auth.functions.http/src/main/java/org/wso2/carbon/identity/conditional/auth/functions/http/CookieFunctionImpl.java [hunk 11]
```diff
@@ -110,7 +110,8 @@ public void setCookie(JsServletResponse response, String name, Object... params)
             Optional.ofNullable((sameSite != null) ? SameSiteCookie.valueOf(sameSite) : null)
                     .ifPresent(cookie::setSameSite);
         }
-        response.addCookie(cookie);
+        //TODO: use response.addCookie(cookie)
```
Comment (senthalan) [inline components/org.wso2.carbon.identity.conditional.auth.functions.http/src/main/java/org/wso2/carbon/identity/conditional/auth/functions/http/CookieFunctionImpl.java:113]: better to remove todo

File: components/org.wso2.carbon.identity.conditional.auth.functions.user/src/main/java/org/wso2/carbon/identity/conditional/auth/functions/user/GetUserSessionsFunctionImpl.java [hunk 12]
```diff
@@ -47,8 +50,13 @@
 
         List<JsUserSession> sessionsForUser = null;
         try {
-            sessionsForUser = getUserSessions(user.getWrapped())
-                    .stream().map(JsUserSession::new).collect(Collectors.toList());
+            if (user instanceof NashornJsAuthenticatedUser) {
```
Comment (senthalan) [inline components/org.wso2.carbon.identity.conditional.auth.functions.user/src/main/java/org/wso2/carbon/identity/conditional/auth/functions/user/GetUserSessionsFunctionImpl.java:53]: can't we remove engine specific classes here?

File: components/org.wso2.carbon.identity.conditional.auth.functions.user/src/main/java/org/wso2/carbon/identity/conditional/auth/functions/user/model/JsApplication.java [hunk 13]
```diff
@@ -1,53 +1,8 @@
-/*
```
Comment (senthalan) [inline components/org.wso2.carbon.identity.conditional.auth.functions.user/src/main/java/org/wso2/carbon/identity/conditional/auth/functions/user/model/JsApplication.java:1]: Any reason to remove the license ?

File: components/org.wso2.carbon.identity.conditional.auth.functions.acr/pom.xml (hunk 1 above)
Comment (shanggeeth) [reply components/org.wso2.carbon.identity.conditional.auth.functions.acr/pom.xml:77]: No, They have been removed

File: components/org.wso2.carbon.identity.conditional.auth.functions.acr/src/main/java/org/wso2/carbon/identity/conditional/auth/functions/acr/SelectAcrFromFunction.java (hunk 2 above)
Comment (shanggeeth) [reply components/org.wso2.carbon.identity.conditional.auth.functions.acr/src/main/java/org/wso2/carbon/identity/conditional/auth/functions/acr/SelectAcrFromFunction.java:1]: Added

File: components/org.wso2.carbon.identity.conditional.auth.functions.acr/src/main/java/org/wso2/carbon/identity/conditional/auth/functions/acr/SelectAcrFromFunction.java (hunk 3 above)
Comment (shanggeeth) [reply components/org.wso2.carbon.identity.conditional.auth.functions.acr/src/main/java/org/wso2/carbon/identity/conditional/auth/functions/acr/SelectAcrFromFunction.java:16]: Added

File: components/org.wso2.carbon.identity.conditional.auth.functions.acr/src/main/java/org/wso2/carbon/identity/conditional/auth/functions/acr/SelectAcrFromFunction.java (hunk 4 above)
Comment (shanggeeth) [reply components/org.wso2.carbon.identity.conditional.auth.functions.acr/src/main/java/org/wso2/carbon/identity/conditional/auth/functions/acr/SelectAcrFromFunction.java:20]: Added

File: components/org.wso2.carbon.identity.conditional.auth.functions.acr/src/main/java/org/wso2/carbon/identity/conditional/auth/functions/acr/SelectAcrFromFunction.java (hunk 5 above)
Comment (shanggeeth) [reply components/org.wso2.carbon.identity.conditional.auth.functions.acr/src/main/java/org/wso2/carbon/identity/conditional/auth/functions/acr/SelectAcrFromFunction.java:72]: That was to keep the number of line characters less 120

File: components/org.wso2.carbon.identity.conditional.auth.functions.analytics/src/main/java/org/wso2/carbon/identity/conditional/auth/functions/analytics/CallAnalyticsFunctionImpl.java (hunk 9 above)
Comment (shanggeeth) [reply components/org.wso2.carbon.identity.conditional.auth.functions.analytics/src/main/java/org/wso2/carbon/identity/conditional/auth/functions/analytics/CallAnalyticsFunctionImpl.java:69]: With Graal Execution original `metadataJsObject` and `payloadDataJsObject` have the scope of their `Context` and become unavailable in the when AsyncProcess` is finally executed and the `Context` would have been closed

File: components/org.wso2.carbon.identity.conditional.auth.functions.analytics/src/test/resources/testng.xml (hunk 10 above)
Comment (shanggeeth) [reply components/org.wso2.carbon.identity.conditional.auth.functions.analytics/src/test/resources/testng.xml:31]: Added

File: components/org.wso2.carbon.identity.conditional.auth.functions.user/src/main/java/org/wso2/carbon/identity/conditional/auth/functions/user/model/JsApplication.java (hunk 13 above)
Comment (shanggeeth) [reply components/org.wso2.carbon.identity.conditional.auth.functions.user/src/main/java/org/wso2/carbon/identity/conditional/auth/functions/user/model/JsApplication.java:1]: No, It's added back

File: components/org.wso2.carbon.identity.conditional.auth.functions.http/src/main/java/org/wso2/carbon/identity/conditional/auth/functions/http/CookieFunctionImpl.java (hunk 11 above)
Comment (shanggeeth) [reply components/org.wso2.carbon.identity.conditional.auth.functions.http/src/main/java/org/wso2/carbon/identity/conditional/auth/functions/http/CookieFunctionImpl.java:113]: Removed

File: components/org.wso2.carbon.identity.conditional.auth.functions.acr/pom.xml [hunk 14]
```diff
@@ -0,0 +1,114 @@
...
+            <plugin>
+                <groupId>org.jacoco</groupId>
+                <artifactId>jacoco-maven-plugin</artifactId>
+            </plugin>
+        </plugins>
+    </build>
+    <properties>
+        <carbon.identity.package.import.version.range>[5.0.0, 6.0.0)</carbon.identity.package.import.version.range>
+    </properties>
+</project>
```
Comment (madurangasiriwardena) [inline components/org.wso2.carbon.identity.conditional.auth.functions.acr/pom.xml:114]: New line.

File: components/org.wso2.carbon.identity.conditional.auth.functions.user/src/test/resources/testng.xml [hunk 15]
```diff
@@ -0,0 +1,34 @@
...
+    </test>
+    <test name="roles-functions-graaljs" parallel="false">
+        <parameter name="scriptEngine" value="graaljs"/>
+        <classes>
+            <class name="org.wso2.carbon.identity.conditional.auth.functions.user.HasAnyOfTheRolesFunctionImplTest"/>
+            <class name="org.wso2.carbon.identity.conditional.auth.functions.user.HasRoleFunctionImplTest"/>
+
+        </classes>
+    </test>
+</suite>
```
[the remark this answers is not in the document]
Comment (shanggeeth) [reply components/org.wso2.carbon.identity.conditional.auth.functions.user/src/test/resources/testng.xml:34]: Added New line here and on similar places

File: components/org.wso2.carbon.identity.conditional.auth.functions.user/src/main/java/org/wso2/carbon/identity/conditional/auth/functions/user/model/JsApplication.java (hunk 13 above)
[the remark this answers is not in the document]
Comment (shanggeeth) [reply components/org.wso2.carbon.identity.conditional.auth.functions.user/src/main/java/org/wso2/carbon/identity/conditional/auth/functions/user/model/JsApplication.java:1]: It was a new class created with an old class name and license was not added. License has now been added.

File: components/org.wso2.carbon.identity.conditional.auth.functions.http/src/test/resources/testng.xml [hunk 16]
```diff
@@ -0,0 +1,31 @@
...
+            <class name="org.wso2.carbon.identity.conditional.auth.functions.http.CookieFunctionImplTest"/>
+        </classes>
+    </test>
+    <test name="cookie-function-graaljs" parallel="false">
+        <parameter name="scriptEngine" value="graaljs"/>
+        <classes>
+            <class name="org.wso2.carbon.identity.conditional.auth.functions.http.CookieFunctionImplTest"/>
+        </classes>
+    </test>
+</suite>
```
[the remark this answers is not in the document]
Comment (shanggeeth) [reply components/org.wso2.carbon.identity.conditional.auth.functions.http/src/test/resources/testng.xml:31]: Added here and on similar places

File: components/org.wso2.carbon.identity.conditional.auth.functions.analytics/src/test/resources/testng.xml (hunk 10 above)
[the remark this answers is not in the document]
Comment (shanggeeth) [reply components/org.wso2.carbon.identity.conditional.auth.functions.analytics/src/test/resources/testng.xml:31]: Added here and on similar instances.

File: components/org.wso2.carbon.identity.conditional.auth.functions.acr/pom.xml (hunk 14 above)
Comment (shanggeeth) [reply components/org.wso2.carbon.identity.conditional.auth.functions.acr/pom.xml:114]: Added

File: components/org.wso2.carbon.identity.conditional.auth.functions.acr/src/test/resources/org/wso2/carbon/identity/conditional/auth/functions/acr/incorrect-function-js-sp-1.xml [hunk 17]
```diff
@@ -0,0 +1,108 @@
+<!--
+  ~ Copyright (c) 2018, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
[the remark this answers is not in the document]
Comment (shanggeeth) [reply components/org.wso2.carbon.identity.conditional.auth.functions.acr/src/test/resources/org/wso2/carbon/identity/conditional/auth/functions/acr/incorrect-function-js-sp-1.xml:2]: Modified to current year

File: components/org.wso2.carbon.identity.conditional.auth.functions.user/src/main/java/org/wso2/carbon/identity/conditional/auth/functions/user/model/nashorn/NashornJsUserSession.java [hunk 18]
```diff
@@ -0,0 +1,67 @@
+/*
+ * Copyright (c) 2019, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
[the remark this answers is not in the document]
Comment (shanggeeth) [reply components/org.wso2.carbon.identity.conditional.auth.functions.user/src/main/java/org/wso2/carbon/identity/conditional/auth/functions/user/model/nashorn/NashornJsUserSession.java:2]: Changed here and o all relevant instances.

**Review by madurangasiriwardena — APPROVED** [review]

**Pull request merged**